includes/delete_datastream.form.inc

```diff
@@ -30,6 +39,33 @@ function islandora_delete_datastream_form(array $form, array &$form_state, Abstr
 }
 
 /**
+ * Return $dsids based on a given DSID and filter.
+ */
+function islandora_datastream_to_purge(AbstractObject $object, $dsid) {
+  module_load_include('inc', 'islandora', 'includes/utilities');
+  $hooks = islandora_invoke_hook_list(ISLANDORA_DERVIATIVE_CREATION_HOOK, $object->models, array($object));
+  $hook_filter = function ($hook_def) use ($dsid) {
+    return isset($hook_def['source_dsid']) && isset($hook_def['destination_dsid']) ?
+      $hook_def['source_dsid'] == $dsid :
+      FALSE;
+  };
+  $hooks = array_filter($hooks, $hook_filter);
+  $dsid_map = function ($hook_definition) {
+    return $hook_definition['destination_dsid'];
+  };
+  $dsids = array_map($dsid_map, $hooks);
+  return $dsids;
```

Comment:
Was [mentioned](https://github.com/Islandora/islandora/pull/586#issuecomment-77172281) by @willtp87, but before returning, we should recurse for each item in `$dsids`, so we also get what derivatives of derivatives there may be... Might look something like:
```php
// -- After the definition of the `$dsid_map` function.
// Start with an empty array.
$dsids = array();
$derived_dsids = array_map($dsid_map, $hooks);
while ($current = array_pop($derived_dsids)) {
  // When we check for derivatives of derivatives, add to the output.
  $dsids[] = $current;
  // Recurse...
  $current_derived = islandora_datastream_to_purge($object, $current);
  // Limit to dsids we haven't already processed or are in the queue.
  $current_diff = array_diff($current_derived, $derived_dsids, $dsids);
  // Add our limited list to those to process.
  $derived_dsids = array_merge($derived_dsids, $current_diff);
}
// Return the flattened "tree" of derivatives.
return $dsids;
```